Rework Fortran configure to allow cross compile overrides

byrnHDF opened the pull request:
Another possibility using defaults and toolchain file could overwrite.
See #5709 and #5718

The expectation is that cross-compiling will use a toolchain file to provide the correct values.
The defaults are only one way to provide defaults for a specific compiler. 

NOTE: The supplied toolchain and Fortran pre-gen files are specific for mingw/gfortran - each platform/compiler will need to supply its own set of files/values.

<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Rework `HDF5UseFortran.cmake` to support cross-compilation by providing default KINDs and sizes.
> 
>   - **Behavior**:
>     - Modify `HDF5UseFortran.cmake` to handle cross-compilation by providing default KINDs and sizes when `CMAKE_CROSSCOMPILING` is true.
>     - Default values for INTEGER and REAL KINDs and their sizes are set when cross-compiling.
>     - Error messages are adjusted to reflect cross-compilation scenarios.
>   - **Logic**:
>     - Conditional checks added for `CMAKE_CROSSCOMPILING` to determine if default values should be used.
>     - Default values include INTEGER KINDs `{1,2,4,8,16}` and REAL KINDs `{4,8,10,16}`.
>     - Default sizes for native kinds are set to 4 for INTEGER and REAL, and 8 for DOUBLE PRECISION when cross-compiling.
>   - **Misc**:
>     - Adjustments to verbose messages to indicate when defaults are used during cross-compilation.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=HDFGroup%2Fhdf5&utm_source=github&utm_medium=referral)<sup> for 0b8f6406ec7f205549ee3de9f912b5f6c1253a91. You can [customize](https://app.ellipsis.dev/HDFGroup/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

byrnHDF committed:
Rework Fortran configure to allow cross compile overrides

byrnHDF commented on config/HDF5UseFortran.cmake:
```diff
@@ -140,83 +140,103 @@ endif ()
+  set (PAC_FC_ALL_INTEGER_KINDS "\{1,2,4,8,16\}" CACHE INTERNAL "Find available INTEGER KINDs for Fortran")
+  set (PAC_FC_ALL_REAL_KINDS "\{4,8,10,16\}" CACHE INTERNAL "Find available REAL KINDs for Fortran")
+  set (${HDF_PREFIX}_PAC_FC_MAX_REAL_PRECISION 33 CACHE INTERNAL "Maximum decimal precision for REALs in Fortran")
+  set (PAC_FORTRAN_NUM_INTEGER_KINDS "5" CACHE INTERNAL "Number of valid integer kinds for Fortran")
```
These are two different items.

byrnHDF commented on config/HDF5UseFortran.cmake:
```diff
@@ -259,128 +279,163 @@ endif()
+else ()
+  # Use the default if there's no cache variable and cross-compiling
+  message (VERBOSE "Find available REAL KINDs for Fortran... yes (cross-compile default)")
+  set (PAC_FC_ALL_REAL_KINDS_SIZEOF "\{4,8,16,16\}" CACHE INTERNAL "Find available REAL KINDs for Fortran")
```
These are correct values for linux

brtnfld replied:
It's not Linux that determines this, but the compiler.  Interpretation of KIND values depends on the compiler. For example, NAG, by default, uses sequential kind numbering, which is not what GNU does.

byrnHDF replied:
Yes, the idea is that toolchain specific overrides will modify this. This is just a simple default - the other option will be to skip the default and require an override.

byrnHDF replied:
I could also change the "(cross-compile default)" to "(cross-compile  gfortran default)"

byrnHDF replied:
The requirement is to specify the correct values in a toolchain file. gfortran defaults are just in case.
The other choice is to halt the config if cross-compile is indicated and no values are provided.

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Add cross-compile template toolchain and use gfortran in defaults

byrnHDF committed:
Add cross-compile action

byrnHDF committed:
install mingw

byrnHDF committed:
correct yaml name

byrnHDF committed:
Fix ext name

byrnHDF committed:
disable plugins until fixed

byrnHDF committed:
correct syntax

byrnHDF committed:
Try setup-mingw

byrnHDF committed:
Try a different toolchain with paths

byrnHDF committed:
fix toolchain

byrnHDF committed:
Try explict install of gfortran

byrnHDF committed:
Disable java

byrnHDF commented:
The mingw cross-compile gets past the configure stage and then chokes on:
FAILED: [code=1] fortran/src/CMakeFiles/H5_buildiface.dir/H5_buildiface.F90.obj 
FAILED: [code=1] fortran/src/CMakeFiles/H5match_types.dir/H5match_types.c.obj

because the values from the defaults are not correctly formatted. Once the syntax is correct then we can update the toolchain with the correct values.

byrnHDF committed:
Adjust default values

byrnHDF committed:
Adjust values and syntax

byrnHDF commented:
Progress has been made on default values - now there is an issue with generating a file as it requires running a Fortran program!
```
[28/3769] Linking Fortran executable bin/H5_buildiface.exe
FAILED: [code=127] bin/H5_buildiface.exe fortran/H5_gen.F90 fortran/static/H5_gen.F90 fortran/shared/H5_gen.F90 /home/runner/work/hdf5/build/fortran/H5_gen.F90 /home/runner/work/hdf5/build/fortran/static/H5_gen.F90 /home/runner/work/hdf5/build/fortran/shared/H5_gen.F90 
: && /usr/bin/x86_64-w64-mingw32-gfortran -fdiagnostics-urls=never -fno-diagnostics-color -O3  fortran/src/CMakeFiles/H5_buildiface.dir/H5_buildiface.F90.obj -o bin/H5_buildiface.exe -Wl,--out-implib,bin/libH5_buildiface.dll.a -Wl,--major-image-version,0,--minor-image-version,0   && cd /home/runner/work/hdf5/build/fortran && wine64 /home/runner/work/hdf5/build/bin/H5_buildiface.exe && cd /home/runner/work/hdf5/build/fortran/static && /usr/local/bin/cmake -E copy_if_different /home/runner/work/hdf5/build/fortran/H5_gen.F90 /home/runner/work/hdf5/build/fortran/static/H5_gen.F90 && cd /home/runner/work/hdf5/build/fortran/shared && /usr/local/bin/cmake -E copy_if_different /home/runner/work/hdf5/build/fortran/H5_gen.F90 /home/runner/work/hdf5/build/fortran/shared/H5_gen.F90
/bin/sh: 1: wine64: not found

```

brtnfld commented:
H5gen.F90 is generated from the standalone code h5_buildiface.F90. It builds all the interfaces for all the kinds/dimensions for the F90 APIs.

byrnHDF commented:
> H5gen.F90 is generated from the standalone code h5_buildiface.F90. It builds all the interfaces for all the kinds/dimensions for the F90 APIs.

So do we provide a way to create a pre-generated file for cross-compiling? And if so, how much of the configure values are needed?

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

brtnfld commented:
> So do we provide a way to create a pre-generated file for cross-compiling? And if so, how much of the configure values are needed?

Currently, we don't, but it uses what it finds during the configuration, so that we could pre-generate it if that information is known.

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Add support for Fortran pregen files

byrnHDF committed:
Update workflow with path to files

byrnHDF committed:
Use full path

byrnHDF commented:
The latest update fixes the Fortran configuration issue with PRE-GENERATED files. And the build succeeds. The tests fail because they are failing with the WINE emulator - maybe someone can suggest a fix?

byrnHDF committed:
CMAKE_CROSSCOMPILING_EMULATOR is handled by CMake added PATH var

byrnHDF commented:
CMake applies the CMAKE_CROSSCOMPILING_EMULATOR for cross compiling and does not need to be specified in the command line. The WINEPATH ENV must be added to calls involving wine emulator. Added this to the mingw toolchain. Only 24 tests failed mostly  h5_test_api and hl_test tests. Will need to investigate which of these tests are expected to fail with wine emulator.

byrnHDF committed:
Correct no cross-compile Fortan custom_commands

byrnHDF committed:
Fix failing tests for mingw

byrnHDF committed:
Correct location for property

byrnHDF commented:
Fixed the h5_api_test by excluding them if cross-compiling.
Fixed the properties for HL and SWMR check
Leaving 5 failures:
         3 - H5TEST-testhdf5-base (Failed)
          8 - H5TESTXPR-cache_api (Failed)
         65 - H5TEST-swmr (Failed)
         89 - H5TEST-filenotclosed (Failed)
         93 - H5TEST-err_compat (Failed)

byrnHDF committed:
Disable HDF5_TEST_API option for cross-compile workflow

byrnHDF committed:
Fix command properties

byrnHDF committed:
Merge develop and fix conflicts

byrnHDF committed:
Merge branch 'develop' into develop-fortran-config

byrnHDF commented:
Note that 3 - H5TEST-testhdf5-base (Failed) is because : 
 HDF_CHECK_TYPE_SIZE (off_t          ${HDF_PREFIX}_SIZEOF_OFF_T) needs to be 4 and not 8

byrnHDF committed:
Fix the HDF5_SIZEOF_OFF_T issue with mingw

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Correct OFF_T size for mingw

byrnHDF committed:
Allow cross-compile workflow to pass with failure testing

byrnHDF commented:
Although the CI doesn't show it - there are only 4 failing tests on my local fedora system. The ubuntu runner seems to have a char set issue causing problems. Likely a missing system dependency?
As far as code changes for CMake, I believe this PR to be complete.

hmaarrfk commented:
16k lines!!!! I really hope this is "improving things" beyond just "cross compilation workflows". I really think cross compilation is critical, but this has definitely gone "beyond" the scope I originally imaged this being.

byrnHDF commented:
So the big change is that CMake will apply the CMAKE_CROSS_COMPILING_EMULATER as needed - had to remove that from the commands
The wine64 emulator requires setting WINEPATH env var. A global setting didn't work for me so I added that to the test properties.
Finally the other changes concern the Fortran configure and the Fortran generated files.

hmaarrfk commented:
Are you able to provide points and tips to the failures in: https://github.com/conda-forge/hdf5-feedstock/pull/268

byrnHDF commented:
> Are you able to provide points and tips to the failures in: [conda-forge/hdf5-feedstock#268](https://github.com/conda-forge/hdf5-feedstock/pull/268)

I haven't tried mpi and cross-compiling - is that the issue?

byrnHDF commented:
Note that the supplied pre-gen files and configure defaults were specific for mingw-gfortran. Other compilers will likely need other values and files. The best thing would be to create a stable of toolchain files and dirs.

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

hmaarrfk commented:
i'm honestly not sure how much time i have to troubleshoot this with you all. i'm really sorry. you can see my checklist, i got pretty far. but i need some hands on keyboard debugging from some1 who is willing to look at things from a conda-forge standpoint. Its free to make a PR, which will run on our infrastructure.

To be honest, it is very likely that we will be able to, by 2026, to run on native runners for OSX on conda-forge. I'm willing to wait, or simply to stick on the HDF5 V1, and "not accelerate" its deployment.

byrnHDF commented:
Yes I believe that running on native machines is the best plan. Our use of native runners has been very beneficial.

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

ajelenak commented:
> it is very likely that we will be able to, by 2026, to run on native runners for OSX on conda-forge.

@hmaarrfk Do you know when in 2026 this may happen?

hmaarrfk commented:
https://github.com/conda-forge/conda-forge.github.io/issues/1781#issuecomment-3243036557

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'develop-fortran-config' of https://github.com/byrnHDF/hdf5.git into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Fix h5dump property command

byrnHDF committed:
install wine in workflow

byrnHDF committed:
revert install wine for workflow

byrnHDF committed:
Try another wine install step

byrnHDF committed:
try another wine install

byrnHDF committed:
Remove wine server option

byrnHDF committed:
Disable tests until wine is properly installed

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Correct name of binary

byrnHDF committed:
Correct binary ext

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge develop changes

byrnHDF committed:
Merge branch 'develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

brtnfld reviewed (dismissed):
📋 Pull Request Summary

  Title: Rework Fortran configuration for cross-compilation supportFocus: Enhancing HDF5UseFortran.cmake to
  handle cross-compilation scenarios by providing sensible default values for Fortran INTEGER and REAL kinds
  when cross-compiling.

  ✅ Strengths

  1. Clear Problem Definition

  - Addresses a specific pain point in cross-compilation workflows
  - Targets mingw/gfortran cross-compilation which is commonly used
  - Provides fallback defaults when runtime detection isn't possible

  2. Modular Implementation

  - Clean conditional logic separating cross-compilation from native builds
  - Maintains backward compatibility with existing build processes
  - Uses CMake best practices for cross-compilation detection

  3. Verbose Configuration

  - Good logging/messaging for configuration tracking
  - Makes it easy to debug configuration issues
  - Clear indication when defaults are being used

  ⚠️ Areas of Concern

  1. Hardcoded Default Values

  # Potential issue: These defaults may not work for all compilers
  set(H5_FORTRAN_DEFAULT_INTEGER_KINDS "1,2,4,8")
  set(H5_FORTRAN_DEFAULT_REAL_KINDS "4,8")

  Risk: Different Fortran compilers may have varying support for these kinds
  Recommendation: Consider compiler-specific detection or documentation

  2. Limited Cross-Compiler Coverage

  - Currently focuses on mingw/gfortran
  - May need expansion for other cross-compilation scenarios (ARM, embedded systems, etc.)
  - Default values might not be appropriate for all target architectures

  3. Testing Gaps

  - Cross-compilation testing is inherently complex
  - Need verification across multiple compiler/target combinations
  - Risk of introducing regressions in native compilation

  🔧 Technical Assessment

  Code Quality: ⭐⭐⭐⭐☆

  - Well-structured CMake code
  - Clear separation of concerns
  - Good error handling and messaging

  Impact Analysis: ⭐⭐⭐⭐⭐

  - Low Risk: Changes are isolated to Fortran configuration
  - High Value: Enables important cross-compilation use cases
  - Backward Compatible: Doesn't affect existing builds

  Security: ⭐⭐⭐⭐⭐

  - No security implications identified
  - Actually reduces risk by providing sane defaults

  📊 Recommendations

  High Priority

  1. Expand Testing
  # Add CI jobs for multiple cross-compilation targets
  - mingw-w64 (current)
  - aarch64-linux-gnu
  - arm-linux-gnueabihf
  2. Document Default Selection
  # Add comments explaining why these specific defaults were chosen
  # Reference: ISO Fortran standard, common compiler behavior, etc.

  Medium Priority

  3. Compiler-Specific Defaults
  # Consider different defaults based on compiler
  if(CMAKE_Fortran_COMPILER_ID MATCHES "GNU")
    set(H5_FORTRAN_DEFAULT_INTEGER_KINDS "1,2,4,8")
  elseif(CMAKE_Fortran_COMPILER_ID MATCHES "Intel")
    # Intel-specific defaults
  endif()
  4. Runtime Validation
  # Add checks to verify the defaults actually work
  # Generate test program to validate kind availability

  Low Priority

  5. Parameterization
  # Allow users to override defaults
  set(HDF5_FORTRAN_CROSS_COMPILE_INTEGER_KINDS "${H5_FORTRAN_DEFAULT_INTEGER_KINDS}"
      CACHE STRING "INTEGER kinds for cross-compilation")

  🎯 Final Recommendation

  APPROVE WITH MINOR REVISIONS ✅

  Rationale:

  - Addresses a legitimate need for cross-compilation support
  - Implementation is clean and low-risk
  - Benefits outweigh the concerns
  - Issues can be addressed in follow-up PRs

  Merge Conditions:

  1. ✅ Add comprehensive cross-compilation testing to CI
  2. ✅ Document the rationale for default kind selections
  3. ✅ Verify defaults work with target compilers
  4. ⚠️ Consider adding compiler-specific logic (future enhancement)

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'develop' into develop-fortran-config

byrnHDF committed:
Add release note

byrnHDF committed:
Merge branch and add release note

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF commented:
> Potential issue: These defaults may not work for all compilers

The requirement is to specify the correct values in a toolchain file. gfortran defaults are just in case.
The other choice is to halt the config if cross-compile is indicated and no values are provided.

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'develop' into develop-fortran-config

byrnHDF committed:
Merge and fix cross-compile changes

byrnHDF committed:
Fix missing space

brtnfld commented on fortran/src/CMakeLists.txt:
```diff
@@ -138,20 +142,62 @@ add_executable (H5match_types
+if (NOT EXISTS "${HDF5_GENERATED_SOURCE_DIR}/H5fortran_types.F90")
+  message (STATUS "No pre-generated H5fortran_types.F90 file found - generating one")
+  # execute the H5match_types program
+  if (CROSSCOMPILING)
```
How is CROSSCOMPILING different from CMAKE_CROSSCOMPILING?

brtnfld replied:
CMAKE_CROSSCOMPILING is the standard variable
Official CMake Documentation:
Source: [CMAKE_CROSSCOMPILING variable documentation](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html)
Definition: "Variable that will be set to true by CMake if the CMAKE_SYSTEM_NAME variable has been manually set (i.e. in a toolchain file or as a cache entry from the cmake command line)."
Evidence from the PR: In [config/toolchain/mingw64.cmake:2](vscode-webview://07jnhl92jfcep6tuigep6h7ts9djin8nc46uphosjp451k2n7hij/config/toolchain/mingw64.cmake#L2):
set (CMAKE_SYSTEM_NAME Windows)
This sets CMAKE_CROSSCOMPILING to TRUE automatically (since the host system is Linux/Ubuntu). The PR uses CROSSCOMPILING (without the CMAKE_ prefix), which is not a standard CMake variable and would need to be manually defined somewhere (which it isn't in this PR).

byrnHDF replied:
> CROSSCOMPILING


Looks like a typo.

byrnHDF replied:
fixed.

brtnfld commented on fortran/src/CMakeLists.txt:
```diff
@@ -138,20 +142,62 @@ add_executable (H5match_types
+        BYPRODUCTS ${HDF5_F90_BINARY_DIR}/H5f90i_gen.h ${HDF5_F90_BINARY_DIR}/H5fortran_types.F90
+        COMMAND    ${CMAKE_COMMAND}
+        ARGS       -E env ${CROSSCOMPILING_PATH}
+        COMMAND    $<TARGET_FILE:H5match_types>
```
This has 2 "COMMAND"  in sequence, 

```
add_custom_command (TARGET H5match_types POST_BUILD
    BYPRODUCTS ${HDF5_F90_BINARY_DIR}/H5f90i_gen.h ${HDF5_F90_BINARY_DIR}/H5fortran_types.F90
    COMMAND ${CMAKE_COMMAND} -E env ${CROSSCOMPILING_PATH} $<TARGET_FILE:H5match_types>
    WORKING_DIRECTORY ${HDF5_F90_BINARY_DIR}
    COMMENT "Generating the H5fortran_types.F90 and H5f90i_gen.h files"
```
)

byrnHDF committed:
Change CROSSCOMPILING to CMAKE_CROSSCOMPILING

byrnHDF committed:
Change CROSSCOMPILING to CMAKE_CROSSCOMPILING

byrnHDF committed:
Change CROSSCOMPILING to CMAKE_CROSSCOMPILING

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Move pregen fortran dir under toolchain

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

lrknox committed:
Merge branch 'develop' into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Fix typo

byrnHDF committed:
Fix missing H5EX conversions

byrnHDF committed:
Fix merge issues

brtnfld commented:
Where is CROSSCOMPILING_PATH defined?

byrnHDF commented on config/toolchain/mingw64.cmake:
```diff
@@ -1,14 +1,37 @@
 set (CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+set (CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
 set (CMAKE_CROSSCOMPILING_EMULATOR wine64)
+set (CROSSCOMPILING_PATH "WINEPATH=/usr/${TOOLCHAIN_PREFIX}/sys-root/mingw/bin/")
```
toolchain defines CROSSCOMPILING_PATH or it must be on the command line.

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config

byrnHDF committed:
Add text file to pregen folder

byrnHDF committed:
Merge branch 'develop-fortran-config' of https://github.com/byrnHDF/hdf5.git into develop-fortran-config

byrnHDF committed:
Merge branch 'HDFGroup:develop' into develop-fortran-config
